src/Index.hpp

```diff
@@ -110,6 +109,77 @@ namespace nbs {
             return types;
         }
 
+
+        // Get timestamp where as many subtypes move forward without moving forward twice.
+        std::uint64_t nextTimestamp(const uint64_t& timestamp,
+                                    const std::vector<TypeSubtype>& types,
+                                    const int& steps) {
+
+            auto matchingIndexItems = this->getIteratorsForTypes(types);  // Iterator for each type(s).
+
+            uint64_t best_timestamp = std::numeric_limits<uint64_t>::max();
+            uint64_t best_delta     = std::numeric_limits<uint64_t>::max();
+            uint64_t earliest_found = std::numeric_limits<uint64_t>::max();
+            uint64_t latest_found   = std::numeric_limits<uint64_t>::min();
+
+            bool bestFound = false;  // If > 1 type, necessary to find earliest or latest timestamp for some edge cases.
+            for (auto& range : matchingIndexItems) {
+                auto& begin = range.first;
+                auto& end   = range.second;
+
+                IndexItemFile target;
+                target.item.timestamp = timestamp;
+
+                if (begin != end) {
+                    // Find the first item with a timestamp greater than the requested timestamp.
+                    auto position =
+                        std::upper_bound(begin, end, target, [](const IndexItemFile& a, const IndexItemFile& b) {
+                            return a.item.timestamp < b.item.timestamp;
+                        });
+
+                    // // Prevent position going past begin or end, return respective iterator timestamp.
+                    // If timestamp precedes start.
+                    if (std::distance(begin, position) + (steps - 1) < 0) {
```

Comment:
So much of this is converting from indices to iterators and back again constantly.

I think it would be much clearer and easier to read if right up the top of this loop after getting position you just did.

```cpp
int length = std::distance(begin, end);
int idx = std::distance(begin, position) - 1;
```

From there you can just do all this logic much more simply
e.g. checking before beginning becomes
`if (idx + steps < 0)`
checking after end becomes
`if (idx + steps >= length)`

Reply:
Refactored.